test/python/pulse/test_builder.py

```diff
@@ -0,0 +1,969 @@
+# that they have been altered from the originals.
+
+"""Test pulse builder context utilities."""
+
+from math import pi
+import unittest
+
+import numpy as np
+
+from qiskit import circuit
+from qiskit import compiler
+from qiskit import pulse
+from qiskit.pulse import builder
+from qiskit.pulse import exceptions
+from qiskit.pulse import macros
+from qiskit.pulse import transforms
+from qiskit.pulse.instructions import directives
+from qiskit.test import QiskitTestCase
+from qiskit.test.mock import FakeOpenPulse2Q
+from qiskit.pulse import pulse_lib, instructions
+
+# pylint: disable=invalid-name
+
+
+class TestBuilder(QiskitTestCase):
+    """Test the pulse builder context."""
+
+    def setUp(self):
+        self.backend = FakeOpenPulse2Q()
+        self.configuration = self.backend.configuration()
+        self.defaults = self.backend.defaults()
+        self.inst_map = self.defaults.instruction_schedule_map
+
+
+class TestContexts(TestBuilder):
+    """Test builder contexts."""
+    def test_align_sequential(self):
+        """Test the sequential alignment context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+
+        with pulse.build() as schedule:
+            with pulse.align_sequential():
+                pulse.delay(3, d0)
+                pulse.delay(5, d1)
+                pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference.insert(0, instructions.Delay(3, d0), mutate=True)
+        reference.insert(8, instructions.Delay(7, d0), mutate=True)
+        # d1
+        reference.insert(3, instructions.Delay(5, d1), mutate=True)
+
+        self.assertEqual(schedule, reference)
+
+    def test_align_left(self):
+        """Test the left alignment context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+        d2 = pulse.DriveChannel(2)
+
+        with pulse.build() as schedule:
+            with pulse.align_left():
+                pulse.delay(11, d2)
+                pulse.delay(3, d0)
+                with pulse.align_left():
+                    pulse.delay(5, d1)
+                    pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference = reference.insert(3, instructions.Delay(5, d1))
+        # d2
+        reference += instructions.Delay(11, d2)
+        self.assertEqual(schedule, reference)
+
+    def test_align_right(self):
+        """Test the right alignment context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+        d2 = pulse.DriveChannel(2)
+
+        with pulse.build() as schedule:
+            with pulse.align_right():
+                pulse.delay(11, d2)
+                pulse.delay(3, d0)
+                with pulse.align_right():
+                    pulse.delay(5, d1)
+                    pulse.delay(7, d0)
```

Comment:
The behavior here would be the same even if it wasn't nested. It should be revised if you want to also test nested align right (the way you did for align left)

Reply:
Excellent Catch! Fixed in 4a47f56b7 with a better implementation of `align_right`.